### PR TITLE
Have tidy ensure that the there are no extra files in the webidls folder #13427

### LINKED ISSUE #13427: Have tidy ensure that the there are no extra files in the webidls folder
We should probably lint that `components/script/dom/webidls` contains only `*.webidl` files.

Code: https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py

### PR DESCRIPTION
**Pull request by gterzian**
Have tidy ensure that the there are no extra files in the webidls folder #13427

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13427 (github issue number if applicable).

- [x] There are tests for these changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13447)

### CODE REVIEW

**Comment by gterzian** [comment]
WIP

**Comment by Wafflespeanut** [comment]
Ping me once you've made the changes :)

**Comment by gterzian** [comment]
@Wafflespeanut @Manishearth ready for review :)

**Comment by gterzian** [comment]
@Wafflespeanut @Manishearth actually I found a bug already, please hold off :)

**Comment by gterzian** [comment]
@Wafflespeanut @Manishearth good for review 👍

**Comment by gterzian** [comment]
@Wafflespeanut @Manishearth any ideas on that build failure? It's a little hard to make out what the failure was...

**Comment by Manishearth** [comment]
Ignore it, servo's travis is broken right now. The first travis build runs tidy and that's enough for now.

**Comment by Ms2ger** [comment]
Should `DIRECTORIES_FOR_FILETYPES` live in the config file?

File: python/tidy/README.rst
Comment (Wafflespeanut) [inline python/tidy/README.rst]: Have a separate commit for the changes to `README`

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Like @Ms2ger said, this should live in the config file (`servo-tidy.toml` in the root dir), and the dict should be obtained from the `config` variable.

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: `only_changed_files` and `exclude_dirs` should be brought in as an argument, same as how we do for other functions.

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: This could be changed to,

``` python
if not any(filename.endswith(ext) for ext in file_extensions):
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: @Ms2ger @Wafflespeanut thank you both for your review, I have implemented the suggested changes, please take another look...

File: servo-tidy.toml
Comment (Wafflespeanut) [inline servo-tidy.toml]: Maybe this should have a different name? (like, `force_ext`?)

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: This could be more generic like,

``` python
"... unexpected extension found for {name}. We only expect {ext} in {dir_name}".format(name=os.path.basename(filename), ext=', '.join(file_extensions), dir_name=os.path.dirname(filename)"
```

Maybe put those stuff in that dictionary before formatting the error?

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Since we're using `get` with defaults, maybe we don't want it in the default config?

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: do you mean removing https://github.com/gterzian/servo/blob/efb85e9e0c70e7e85b9d74ba54782c1e98c139cb/python/tidy/servo_tidy/tidy.py#L39 ?

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: good one 👍

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Yes, since we're getting `{}` by default anyway :)

File: servo-tidy.toml
Comment (Wafflespeanut) [inline servo-tidy.toml]: No, it may not be the best, I just thought we could keep the table names as precise as possible :)

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: ok 👍

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: By the way, the `get` with default `{}` is when accessing the parsed `toml` file (see(this line)[https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py#L712]), 
as opposed to the python `config` object that is (declared here)[https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py#L27]

So if I remove the `check_ext` key from the default config(the python one), then at some point in `parse_config` I will have to do a `config['check_ext'] = {}`. 

Perhaps actually 'declaring' this key at the top of the page, is more explicit than doing so somewhere in code on the page?

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: I also have another small problem :)

Currently I pass the directories to check as an argument to the `check_directory_files` function, inside the main `scan` function, by actually grabbing from the config([see](https://github.com/gterzian/servo/blob/check_webidls_folder_files/python/tidy/servo_tidy/tidy.py#L855)). This makes testing that function very easy, as in the test I just pass another list of directories ([see](https://github.com/gterzian/servo/blob/check_webidls_folder_files/python/tidy/servo_tidy_tests/test_tidy.py#L39))

The funny thing is that, at this point in the scan function, the config hasn't been parsed yet. The parsing only occurs later, during the first iteration of the generators :) 

So if I remove the default key on top, I get a KeyError at that point.

 If I keep the default key, it works, because the initially empty dict is later mutated during parsing, so by the time the exectution gets to [here](https://github.com/gterzian/servo/blob/check_webidls_folder_files/python/tidy/servo_tidy/tidy.py#L743) we have the value we expect. 

So the question is, should I move the access to `config` to inside `check_directory_files`? If I do, it makes for harder testing of that function, since I would have to mock the config object.

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: OK I have found a solution, I think, I'll push the code and we'll be able to discuss in more concrete way :)

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy_tests/test_tidy.py]: Why can't we just `collect_errors_for_files` like we do in the other test methods? That way, we won't have to rely on the config. We could just assert that we deny non-webidl files in test folders.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by Wafflespeanut** [comment]
I'm sorry it took me a while to look into the changes :)

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy_tests/test_tidy.py]: It's a good idea, and the problem is that `collect_errors_for_files` is really written for the 'standard checks' part, which checks individual files. It would require changing quite a few things to plug this "directory checks" part into it.

For example:

* In the main `scan` function, `collect_errors_for_files` only receives files that have some limited subset of expected extensions, while in our case we actually want to check for the absence of any type of unexpected files. 

* We could call `collect_errors_for_files` separately just for our part of the workflow, however `collect_errors_for_files` contains all sorts of stuff that makes it unusable in our case, such as [specific errors for empty files](https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py#L746) and [printing an inappropriate message in our case](https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py#L738).

Right now we are already re-using the `get_file_list`, which is more generic, and I feel like the `collect_errors_for_files` is not generic, rather written specificaly for the other checks. 

I'd say it's easier to not try to re-use it, and keep the "directory checks" separate from the other "file checks". 


For example, that part of the workflow only checks files with certain extensions, while in this case we really want to check for the presence of any type of unexpected files.  There is also some other small stuff, for example this line:

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: Ok, that's done :)

File: python/tidy/README.rst
Comment (gterzian) [inline python/tidy/README.rst]: thanks, I've removed it :)

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Nit: Let's be a bit more clear: `"We only expect files with {ext} extensions in {dir_name}"`

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy_tests/test_tidy.py]: Yeah, sorry, I'd totally forgotten that we're actually checking directories :smile:

**Review by ghost — APPROVED** [review]
That said, we could train the function to handle directories, but that shouldn't be a part of this PR, as we'll probably need to refactor tidy along the way. Right now, this PR is good to go (with the nit addressed)

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: Ok, done 😄

**Comment by Wafflespeanut** [comment]
Thanks for doing this!

@bors-servo r+

**Comment by Wafflespeanut** [comment]
@bors-servo r-

Turns out that we need to install `mock` library in Windows.

    Traceback (most recent call last):
      File "mach", line 25, in <module>
        main(sys.argv)
      File "mach", line 21, in main
        mach = mach_bootstrap.bootstrap(topdir)
      File "C:\projects\servo\python\mach_bootstrap.py", line 223, in bootstrap
        mach.load_commands_from_file(os.path.join(topdir, path))
      File "C:\projects\servo\python\_virtualenv\Lib\site-packages\mach\main.py", line 265, in load_commands_from_file
        imp.load_source(module_name, path)
      File "C:\projects\servo\python\servo\testing_commands.py", line 32, in <module>
        from servo_tidy_tests import test_tidy
      File "C:\projects\servo\python\tidy\servo_tidy_tests\test_tidy.py", line 12, in <module>
        import mock
    ImportError: No module named mock
    Command exited with code 1

Can we get around this without using `mock` library?

**Comment by Wafflespeanut** [comment]
Okay, I'm thinking of improving this a bit more!

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: `get_file_list` is useful for traversing through a folder structure. But, we want to make sure that a directory contains only 'files' of certain extensions. I think `os.listdir` would be simpler and would fit best for our purpose. That way, we can also enforce that a directory only contains the desired files, and doesn't contain any sub-dirs. And, we don't want the "ignore" part, since we won't be having any sub-dirs in these dirs.

**Comment by gterzian** [comment]
@Wafflespeanut The mock library is already part of requirements.txt, I actually didn't have to install it on my mac. What is the problem with Windows?

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: As we use `os.listdir` (as said), we should make this take a `dict` which is hopefully config['check_ext'].

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy_tests/test_tidy.py]: And, we won't be needing the `mock` library to patch the config along the way. We could just pass a custom `dict` to the function.

**Review by ghost — CHANGES_REQUESTED** [review]
As for Windows, I'm not sure. It's the log from appveyor. If that's a dependency on Windows, then we don't probably wanna go for it just for the sake of testing this function.

And, sorry I didn't take a keen look at your PR earlier :)

**Comment by gterzian** [comment]
@Wafflespeanut actually when I wrote earlier that it's already part of requirements.txt, I think I was wrong, perhaps I was looking at the requirements.txt of another project I was looking at.

yet the test passed on my mac, not sure how actually...

**Comment by Wafflespeanut** [comment]
I can import `mock` in my machine too. It's just that it appears to be a dependency on Windows. So, let's opt for a different way :)

**Comment by gterzian** [comment]
@Wafflespeanut ok so yes, we can go back to passing the list of dirs at function call, since we have the default key in the config it will work (earlier the problem was that without that default key, trying to access it in the `scan` function would throw a `KeyError`, because the config only get's parsed later, during the first iteration of the entire "generator pipeline")

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy_tests/test_tidy.py]: Ok, thanks again for the suggestions, all implemented... 😄

**Comment by gterzian** [comment]
@Wafflespeanut one sec, forgot to remove the import...

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: We should make sure that there are only files. There's an `os.path.isfile` for this check

File: python/tidy/servo_tidy_tests/test_tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy_tests/test_tidy.py]: This should all be `os.path.join(base_path, test_rel_path)`

Comment (gterzian) [inline python/tidy/servo_tidy_tests/test_tidy.py]: what do you mean?

Comment (Wafflespeanut) [inline python/tidy/servo_tidy_tests/test_tidy.py]: There's a `base_path` on the top of the file. We should make use of it and `os.path.join` with the relative paths where your test files reside.

Comment (gterzian) [inline python/tidy/servo_tidy_tests/test_tidy.py]: ah ok

File: python/tidy/servo_tidy/tidy.py
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Um, shouldn't we be throwing an error here? Or else, there wouldn't be any reason to do this check (since directories won't have the extension suffix anyway)

**Review by ghost — APPROVED** [review]

**Comment by Wafflespeanut** [comment]
Once the final comment is addressed, we can land this :)

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: yes, that might be a good idea. Actually I was wondering if you meant: "making sure we're only checking extensions for actual files", or rather "besides checking file extensions, let's also check if those are files in the first place". 

If it's not a file (I checked the docs and symlinks are considered files for this function, to be honest I do not know what would be such a non-file 😄 ), should we really throw an error? Or do you mean yielding an error like is done with incorrect extensions? If so, what should be the error message? Just the same as for wrong extensions?

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: ah, could it be that a sub directory would be such a non-file case? In which case we could yield an error saying there shouldn't be a sub-dir?

Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: Never mind. I thought we could differentiate directories and files, but that would be unnecessary complication.

Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: yep, looks like a subdir would fall under that case, ok I'll handle that one specifically then.

Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: No, we don't need to be so strict. Let's revert that check, and we'll land this :)

**Comment by Wafflespeanut** [comment]
Also, Travis shows an error...

    Traceback (most recent call last):

      File "/home/travis/build/servo/servo/python/tidy/servo_tidy_tests/test_tidy.py", line 44, in test_directory_checks

        self.assertNoMoreErrors(errors)

      File "/home/travis/build/servo/servo/python/tidy/servo_tidy_tests/test_tidy.py", line 25, in assertNoMoreErrors

        errors.next()

      File "/home/travis/build/servo/servo/python/tidy/servo_tidy/tidy.py", line 745, in check_directory_files

        files = os.listdir(directory)

    OSError: [Errno 2] No such file or directory: 'python/tidy/servo_tidy_tests/dir_check/empty'

File: python/tidy/servo_tidy/tidy.py
Comment (gterzian) [inline python/tidy/servo_tidy/tidy.py]: Ok so I removed the entire `isfile` check, if necessary we could always add a 'allow_subdir' option or something, for now we could just keep it focused on checking file extensions and assume the dirs will not have sub-dirs(I checked and now it actually sees a sub-dir as a file with the wrong extension...).

I also removed the test case using the `empty` dir, which off course never got added to the repo since it's empty and we don't need it anyway, if the dir is empty the function will simply not check anything since the iterable it operates on will be empty too...

**Comment by Wafflespeanut** [comment]
Thanks for sticking with this! Sorry if I was confusing/nitpicky :smile: 

@bors-servo r+

**Comment by gterzian** [comment]
@Wafflespeanut Not at all, thanks a lot for all the feedback, it was very helpful! 👍

**Comment by jdm** [comment]
```
======================================================================
FAIL: test_directory_checks (servo_tidy_tests.test_tidy.CheckTidiness)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/servo/buildbot/slave/linux-dev/build/python/tidy/servo_tidy_tests/test_tidy.py", line 41, in test_directory_checks
    self.assertEqual("Unexpected extension found for test.rs. We only expect files with webidl, test extensions in {0}".format(error_dir), errors.next()[2])
AssertionError: 'Unexpected extension found for test.rs. We only expect files with webidl, test extensions in python/tidy/servo_tidy_tests/dir_check/webidl_plus' != 'Unexpected extension found for test2.rs. We only expect files with webidl, test extensions in python/tidy/servo_tidy_tests/dir_check/webidl_plus'

----------------------------------------------------------------------
Ran 19 tests in 0.011s

FAILED (failures=1)
```

**Comment by gterzian** [comment]
@Wafflespeanut @jdm Ok, I think this was due to the fact that the order of filenames returned by `os.listdir` is not necessarily the same, so I am ordering the names now. That should fix it.

**Comment by Wafflespeanut** [comment]
I'm surprised, because Travis didn't show a failure on that earlier.

@bors-servo r+

**Comment by jdm** [comment]
@bors-servo: retry
* #13501

**Pull request merged**
